src/main/java/hudson/remoting/Launcher.java

```diff
@@ -456,25 +480,146 @@ private SSLSocketFactory getSSLSocketFactory()
+        }
+
+        // Parse the server's pseudo-JNLP output
+        Launcher bootstrap = new Launcher();
+        CmdLineParser parser = new CmdLineParser(bootstrap);
+        parser.parseArgument(jnlpArgs.toArray(new String[0]));
+        normalizeArguments(bootstrap);
+        validateInboundAgentArgs(bootstrap);
+
+        // Apply the results
+        assert urls.isEmpty();
+        urls.addAll(bootstrap.urls);
+        if (bootstrap.secret != null) {
+            secret = bootstrap.secret;
+        }
+        if (bootstrap.name != null) {
+            name = bootstrap.name;
+        }
+        if (bootstrap.webSocket) {
+            webSocket = true;
+        }
+        if (bootstrap.tunnel != null) {
+            tunnel = bootstrap.tunnel;
+        }
+        if (bootstrap.workDir != null) {
+            workDir = bootstrap.workDir;
+        }
+        if (!WorkDirManager.DirType.INTERNAL_DIR.getDefaultLocation().equals(bootstrap.internalDir)) {
+            internalDir = bootstrap.internalDir;
+        }
+        if (bootstrap.failIfWorkDirIsMissing != WorkDirManager.DEFAULT_FAIL_IF_WORKDIR_IS_MISSING) {
+            failIfWorkDirIsMissing = bootstrap.failIfWorkDirIsMissing;
+        }
+    }
+
+    private static void normalizeArguments(Launcher launcher) throws CmdLineException {
+        if (!launcher.args.isEmpty()) {
+            if (launcher.args.size() != 2) {
+                throw new CmdLineException(null, "Two arguments required, but got " + launcher.args);
+            }
+            if (launcher.secret == null) {
+                launcher.secret = launcher.args.get(0);
+            } else {
+                throw new CmdLineException(null, "Cannot provide secret via both named and positional arguments");
+            }
+            if (launcher.name == null) {
+                launcher.name = launcher.args.get(1);
+            } else {
+                throw new CmdLineException(null, "Cannot provide name via both named and positional arguments");
+            }
+            launcher.args.clear();
+        }
+    }
+
+    private static void validateInboundAgentArgs(Launcher launcher) throws CmdLineException {
+        assert launcher.args.isEmpty() : "should have been normalized previously";
+        if (launcher.secret == null) {
+            throw new CmdLineException(null, "Secret is required for inbound agents");
+        }
+        if (launcher.name == null) {
+            throw new CmdLineException(null, "Name is required for inbound agents");
+        }
+        if (launcher.urls.isEmpty() && launcher.directConnection == null) {
+            throw new CmdLineException(null, "At least one URL is required for inbound agents");
+        }
+        if (launcher.webSocket) {
+            assert !launcher.urls.isEmpty(); // depends = "-url"
+            if (launcher.urls.size() > 1) {
+                throw new CmdLineException(null, "Only a single URL is supported for WebSocket agents");
+            }
+        }
+    }
```

Comment:
Invariant: whenever connecting as an inbound agent, invoke this method to validate the arguments. This happens once in the normal flow and twice in the `-jnlpUrl` flow.